x/superfluid/keeper/migrate_test.go

```diff
@@ -0,0 +1,211 @@
+				{BalancerPoolId: pool.GetId(), ClPoolId: clPool.GetId()},
+			}}
+			gammKeeper.SetMigrationInfo(ctx, migrationRecord)
+
+			// The amount of coins we lock is the amount of LP tokens the account that joined the pool has.
+			coinsToLock := poolShareOut
+
+			// The unbonding duration is the same as the staking module's unbonding duration.
+			unbondingDuration := stakingKeeper.GetParams(ctx).UnbondingTime
+
+			// Lock the LP tokens for the duration of the unbonding period.
+			lockID := suite.LockTokens(poolJoinAcc, sdk.NewCoins(coinsToLock), unbondingDuration)
+
+			// Superfluid delegate the lock if the test case requires it.
+			// Note the intermediary account that was created.
+			intermediaryAcc := types.SuperfluidIntermediaryAccount{}
+			if tc.superfluidDelegated {
+				err = superfluidKeeper.SuperfluidDelegate(ctx, poolJoinAcc.String(), lockID, valAddr.String())
+				suite.Require().NoError(err)
+				intermediaryAccConnection := superfluidKeeper.GetLockIdIntermediaryAccountConnection(ctx, lockID)
+				intermediaryAcc = superfluidKeeper.GetIntermediaryAccount(ctx, intermediaryAccConnection)
+			}
+
+			// Superfluid undelegate the lock if the test case requires it.
+			if tc.superfluidUndelegating {
+				err = superfluidKeeper.SuperfluidUndelegate(ctx, poolJoinAcc.String(), lockID)
+				suite.Require().NoError(err)
+			}
+
+			// Unlock the lock if the test case requires it.
+			if tc.unlocking {
+				// If lock was superfluid staked, we can't unlock via `BeginUnlock`,
+				// we need to unlock lock via `SuperfluidUnbondLock`
+				if tc.superfluidUndelegating {
+					err = superfluidKeeper.SuperfluidUnbondLock(ctx, lockID, poolJoinAcc.String())
+					suite.Require().NoError(err)
+				} else {
+					lock, err := lockupKeeper.GetLockByID(ctx, lockID)
+					suite.Require().NoError(err)
+					err = lockupKeeper.BeginUnlock(ctx, lockID, lock.Coins)
+					suite.Require().NoError(err)
+
+					// add time to current time to test lock end time
+					ctx = ctx.WithBlockTime(ctx.BlockTime().Add(time.Hour * 24))
+				}
+			}
+
+			lock, err := lockupKeeper.GetLockByID(ctx, lockID)
+			suite.Require().NoError(err)
+
+			// Run the unlock and migrate logic.
+			amount0, amount1, _, _, _, err := superfluidKeeper.UnlockAndMigrate(ctx, poolJoinAcc, lockID, poolShareOut)
+			suite.Require().NoError(err)
+
+			suite.AssertEventEmitted(ctx, gammtypes.TypeEvtPoolExited, 1)
+
+			// exitPool has rounding difference.
+			// We test if correct amt has been exited and frozen by comparing with rounding tolerance.
+			defaultErrorTolerance := osmomath.ErrTolerance{
+				AdditiveTolerance: sdk.NewDec(2),
+				RoundingDir:       osmomath.RoundDown,
+			}
+			suite.Require().Equal(0, defaultErrorTolerance.Compare(joinPoolAmt.AmountOf("foo"), amount0))
+			suite.Require().Equal(0, defaultErrorTolerance.Compare(joinPoolAmt.AmountOf("stake"), amount1))
+
```

Comment:
Can we add a check that checks if the frozen unitl has been properly migrated?

Reply:
What I ended up doing was checking that the expected position with a frozenUntil value was created https://github.com/osmosis-labs/osmosis/pull/4129/commits/5f0d2a048e2e2df35e8866eabf91cbcbf24210e6

We test `getExistingLockRemainingDuration` on its own, so we have a guarantee that whatever that frozenUntil value that gets returned is valid. Your point still stands though and it is now indirectly checked!